Zion Smith: 49280339
Izzah Kayani: 49249154

# Preperation and Overview

### Buisiness Use

Since the COVID-19 pandemic, working remotely has exploded in popularity in the workforce. What started off as a forced pivot due to a gloabl pandemic has now turned into the perferred work style for thousands of employees. Identifying the long term effects of working remotely is very important for companies who are experiencing the shift towards working from home. This dataset holds information about 150 remote workers, tracked over 90 days. The attributes are sepereated into two categories: Behavioral and Psychological. Behavioral attributes were gathered by a background python processes, and the psychological gathered via daily surveys and the UCLA loneliness scale. Each attribute is used to find three classes, burnout risk, which can be low, medium and high. The end goal of this dataset is to be able to predict burnout ahead of time, so that it can be subdued before it occurs. The results can be very useful for companies that have remote employees to improve retention, prevent burnout, and prevent productivity loss.

How well the algorithm needs to perform: According to the dataset source, each employee burnout costs  7,000− 13,000 dollars due to the loss of productivity. The source also states that the algorithim is expected to have 80-86% accuracy. Lets say a company has 150 employees, with 40 of them medium-high burnout risk. At 80% accuracy, the algorithim would be able to predict 32 employees that are medium-high risk. Based on the burnout costs, the company could potentially save  224,000− 442,000 dollars by intervening early. This means that the algorithim does not need to be 100% accurate to be useful, 80% is enough to make a positive impact.

In [ ]:
import pandas as pd
import numpy as np

print('Pandas:', pd.__version__)
print('Numpy:',np.__version__)

df = pd.read_csv('/content/wfh_burnout_dataset.csv') # read in the csv file

pd.set_option('display.max_columns', None)
df['burnout_risk'] = df['burnout_risk'].replace(['Low', 'Medium', 'High'], [1,2,3])

df= df.drop(columns='user_id', axis=1)

tmp_df = pd.get_dummies(df['day_type'], prefix='day_type')
df = pd.concat((df, tmp_df), axis=1)
del df['day_type']

df


Pandas: 2.2.2
Numpy: 2.0.2


/tmp/ipykernel_282/1391530920.py:10: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['burnout_risk'] = df['burnout_risk'].replace(['Low', 'Medium', 'High'], [1,2,3])


,work_hours,screen_time_hours,meetings_count,breaks_taken,after_hours_work,app_switches,sleep_hours,task_completion,isolation_index,fatigue_score,burnout_score,burnout_risk,day_type_Weekday,day_type_Weekend
0,7.74,5.99,3,7,0,50,7.50,74.80,4,3.51,17.23,1,True,False
1,8.48,7.23,3,5,0,63,7.31,86.36,4,5.15,24.01,1,True,False
2,8.39,7.14,1,5,0,27,6.73,89.23,8,7.85,51.46,2,False,True
3,12.48,12.23,4,1,1,89,6.55,64.47,5,10.00,55.50,2,True,False
4,8.85,7.35,4,6,0,40,8.03,98.62,3,3.45,11.95,1,False,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,7.93,6.43,1,6,0,22,7.75,73.67,4,2.57,13.16,1,False,True
1996,8.37,6.87,5,6,0,89,7.00,80.05,5,7.13,35.41,2,False,True
1997,10.97,10.47,13,2,0,93,5.86,46.95,6,10.00,58.79,2,False,True
1998,8.48,6.98,0,6,0,29,8.12,72.57,6,6.41,32.64,1,True,False


Burnout Risk was originally Low, Medium, High. This target was replaced by 1, 2, 3 representing Low, Medium, High respectively. This was done because it is a discrete Ordinal value. Column "user_id" was removed because it does not have any relevancy to the burnout risk score, it simply identifies which worker the data is about. We also changed the day_type variables to two variables, day_type_weekend and day_type_weekday. The values of these two variables can be true or false. This allows the machine learning model to process the boolean values without implying ordinal relationships.

In [ ]:
df.describe(include='all')

,work_hours,screen_time_hours,meetings_count,breaks_taken,after_hours_work,app_switches,sleep_hours,task_completion,isolation_index,fatigue_score,burnout_score,burnout_risk,day_type_Weekday,day_type_Weekend
count,2000.000000,2000.000000,2000.000000,2000.000000,2000.00000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000.000000,2000,2000
unique,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2,2
top,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,True,False
freq,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1435,1435
mean,8.955995,7.689495,3.728500,5.066000,0.15650,56.025500,6.894900,80.090670,4.875500,6.560825,35.656590,1.559500,NaN,NaN
std,1.724214,1.983181,2.874877,1.762441,0.36342,23.702463,0.983893,12.708583,1.656011,2.218042,17.315039,0.620203,NaN,NaN
min,4.730000,2.910000,0.000000,0.000000,0.00000,5.000000,3.580000,40.000000,3.000000,1.450000,6.060000,1.000000,NaN,NaN
25%,7.750000,6.280000,2.000000,4.000000,0.00000,39.000000,6.320000,71.672500,3.000000,4.740000,21.005000,1.000000,NaN,NaN
50%,8.535000,7.290000,3.000000,5.000000,0.00000,52.000000,7.080000,80.900000,5.000000,6.295000,32.330000,1.000000,NaN,NaN
75%,9.980000,8.882500,5.000000,6.000000,0.00000,71.000000,7.570000,89.650000,6.000000,8.480000,47.475000,2.000000,NaN,NaN


In [ ]:
df_grouped = df.groupby(by='burnout_risk')

for val,grp in df_grouped:
    print(f'There are {len(grp)} people that are classified as {val} class.')

There are 1019 people that are classified as 1 class.
There are 843 people that are classified as 2 class.
There are 138 people that are classified as 3 class.


In [ ]:
from sklearn.model_selection import train_test_split
import numpy as np
import plotly.graph_objects as go

y = df['burnout_risk'].to_numpy()
X = df.drop(columns='burnout_risk').to_numpy().astype(float)

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, test_size=0.2, random_state=42)
print("Training set shape:", X_train.shape, y_train.shape)
print("Testing set shape:", X_test.shape, y_test.shape)


Training set shape: (1600, 13) (1600,)
Testing set shape: (400, 13) (400,)


In [ ]:
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestClassifier

# Initialize a classifier (e.g., Random Forest)
clf = RandomForestClassifier(random_state=42)

# Perform 5-fold cross-validation on the training set
cv_scores = cross_val_score(clf, X_train, y_train, cv=5, scoring='accuracy')

# Display cross-validation results
print("Cross-validation scores:", cv_scores)
print("Mean cross-validation accuracy:", cv_scores.mean())

In [ ]:
import numpy as np

# Assuming y is your target variable
num_classes = len(np.unique(y))
print(f"Number of classes: {num_classes}")
print(f"Unique class labels: {np.unique(y)}")

80/20 split will work for this dataset because of its moderate size (2,000) samples and the cross validation model shows that the 80/20 split is working properly for this dataset. However, there is an imbalance in the classes, as most of the data falls in to 1 and 2, but not much in 3. This shows that while 80/20 may work fine, there may be a better way to split the data, like stratified sampling.

#Modeling

##Binary Logistic Regression

In [ ]:
import numpy as np
from scipy.optimize import minimize
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from scipy.special import expit
from numpy.linalg import pinv


class CustomBinaryLogisticRegressionBase:
    def __init__(self, eta, solver='steepest_ascent', max_iter=20, C=1.0, tol=1e-4, mini_batch_size=32, regularization='none'):
        self.eta = eta
        self.max_iter = max_iter
        self.C = C
        self.solver = solver
        self.tol = tol
        self.mini_batch_size = mini_batch_size
        self.regularization = regularization
        self.classifiers_ = []

    def __str__(self):
        return 'Base Binary Logistic Regression Object, Not Trainable'

    # convenience, private and static:
    @staticmethod
    def _sigmoid(theta):
        # increase stability, redefine sigmoid operation
        return expit(theta) #1/(1+np.exp(-theta))

    @staticmethod
    def _add_intercept(X):
        return np.hstack((np.ones((X.shape[0],1)),X)) # add bias term

    # public:
    def predict_proba(self, X, add_intercept=True):
        # add bias term if requested
        Xb = self._add_intercept(X) if add_intercept else X
        return self._sigmoid(Xb @ self.weights) # return the probability y=1

    def predict(self, X):
        return (self.predict_proba(X) > 0.5) #return the actual prediction

    def gradient(self, X, y):
        # vectorized gradient: (1/m) * X.T @ (y - h)
        m = len(y)
        h = self._sigmoid(X @ self.weights)
        grad = (1 / m) * (X.T @ (y - h))

        if self.regularization == 'l1':
            reg = (self.C / m) * np.sign(self.weights)
            reg[0] = 0  # do not regularize bias
            grad -= reg
        elif self.regularization == 'l2':
            reg = (self.C / m) * self.weights
            reg[0] = 0  # do not regularize bias
            grad -= reg
        elif self.regularization == 'both':
            reg = (self.C / m) * np.sign(self.weights)
            reg[0] = 0  # do not regularize bias
            grad -= reg
            reg = (self.C / m) * self.weights
            reg[0] = 0  # do not regularize bias
            grad -= reg
        else:
            reg = np.zeros(self.weights.shape)

        return grad

    def fit(self, X, y):
        Xb = self._add_intercept(X) # add bias term
        num_samples, num_features = Xb.shape
        self.weights = np.zeros(num_features, dtype=float)

        if self.solver == 'steepest_ascent':
            self.steepest_ascent(Xb, y)
        elif self.solver == 'newton':
            self.newton(Xb, y)
        elif self.solver == 'stochastic_gradient_ascent':
            self.stochastic_gradient_ascent(Xb, y)

    def steepest_ascent(self, X, y): # Optimize with steepest ascent
        for _ in range(self.max_iter):
            grad = self.gradient(X, y)
            self.weights += self.eta * grad
            if np.max(np.abs(grad)) < self.tol:
                break

    def stochastic_gradient_ascent(self, X, y):
        for _ in range(self.max_iter):
            idxs = np.random.choice(len(y), self.mini_batch_size)
            grad = self.gradient(X[idxs], y[idxs])
            self.weights += self.eta * grad
            if np.max(np.abs(grad)) < self.tol:
                break

    def newton(self, X, y):
        for _ in range(self.max_iter):
            g = self.predict_proba(X, add_intercept=False).ravel() # get sigmoid value for all classes
            hessian = X.T @ np.diag(g*(1-g)) @ X # calculate the hessian
            if self.regularization in ['l2', 'both']:
                hessian -= 2 * self.C

            grad = self.gradient(X, y)
            self.weights += (pinv(hessian) @ grad)
            if np.max(np.abs(grad)) < self.tol:
                break


blr = CustomBinaryLogisticRegressionBase(0.1)
print(blr)

Base Binary Logistic Regression Object, Not Trainable


## MultiClass Logistic Regression: One versus All

In [ ]:
class MultiClassLogisticRegression:
    def __init__(self, eta, solver='steepest_ascent', max_iter=20, C=1.0, tol=1e-4, mini_batch_size=32, regularization='none'):
        self.eta = eta
        self.max_iter = max_iter
        self.C = C
        self.solver = solver
        self.tol = tol
        self.mini_batch_size = mini_batch_size
        self.regularization = regularization
        self.classifiers_ = []

    def __str__(self):
        if(hasattr(self,'w_')):
            return 'MultiClass Logistic Regression Object with coefficients:\n'+ str(self.w_) # is we have trained the object
        else:
            return 'Untrained MultiClass Logistic Regression Object'

    @property
    def coef_(self):
        if(hasattr(self,'w_')):
            return self.w_[:,1:]
        else:
            return None

    @property
    def intercept_(self):
        if(hasattr(self,'w_')):
            return self.w_[:,0]
        else:
            return None

    def fit(self,X,y):
        num_samples, num_features = X.shape
        self.unique_ = np.unique(y) # get each unique class value
        num_unique_classes = len(self.unique_)
        self.classifiers_ = [] # will fill this array with binary classifiers

        for i,yval in enumerate(self.unique_): # for each unique value
            y_binary = (y==yval) # create a binary problem
            # train the binary classifier for this class
            blr = CustomBinaryLogisticRegressionBase(eta=self.eta, solver=self.solver,
                                          max_iter=self.max_iter, C=self.C,
                                          tol=self.tol, mini_batch_size=self.mini_batch_size,
                                          regularization=self.regularization)
            blr.fit(X,y_binary)
            # add the trained classifier to the list
            self.classifiers_.append(blr)

        # save all the weights into one matrix, separate column for each class
        self.w_ = np.hstack([x.weights for x in self.classifiers_]).T

    def predict_proba(self,X):
        probs = []
        for blr in self.classifiers_:
            probs.append(blr.predict_proba(X).reshape(-1,1)) # get probability for each classifier

        return np.hstack(probs) # make into single matrix

    def predict(self,X):
        return self.unique_[np.argmax(self.predict_proba(X),axis=1)] # take argmax along row

lr = MultiClassLogisticRegression(eta=0.1, max_iter=1500)
print(lr)

Untrained MultiClass Logistic Regression Object


##Training
Now let's find the best optimization technique and regularization term for the dataset.

Let's start with an initial run

In [ ]:
lr = MultiClassLogisticRegression(eta=0.1, max_iter=500)
lr.fit(X_train, y_train)
yhat = lr.predict(X_test)
print('Accuracy:', accuracy_score(y_test, yhat))


Accuracy: 0.8225


Now let's test with different C and solver values

In [ ]:
import time

solvers = ['steepest_ascent', 'stochastic_gradient_ascent', 'newton']
C_vals = [0.001, 0.01, 0.1, 1.0, 10.0]

results = []

for solver in solvers:
    for C in C_vals:
      lr = MultiClassLogisticRegression(eta=0.1, max_iter=500, solver=solver, C=C, regularization='l2')
      start = time.time()
      lr.fit(X_train, y_train)
      train_time = time.time() - start
      yhat = lr.predict(X_test)
      accuracy = accuracy_score(y_test, yhat)
      results.append({'solver': solver, 'C': C, 'accuracy': accuracy, 'training_time': train_time})

      print(f'Solver: {solver}, C: {C}, Accuracy: {accuracy}, Time: {train_time:.4f}s')


Solver: steepest_ascent, C: 0.001, Accuracy: 0.83, Time: 0.5411s
Solver: steepest_ascent, C: 0.01, Accuracy: 0.89, Time: 0.2336s
Solver: steepest_ascent, C: 0.1, Accuracy: 0.83, Time: 0.0824s
Solver: steepest_ascent, C: 1.0, Accuracy: 0.8925, Time: 0.0877s
Solver: steepest_ascent, C: 10.0, Accuracy: 0.87, Time: 0.0957s
Solver: stochastic_gradient_ascent, C: 0.001, Accuracy: 0.6975, Time: 0.0475s
Solver: stochastic_gradient_ascent, C: 0.01, Accuracy: 0.64, Time: 0.0745s
Solver: stochastic_gradient_ascent, C: 0.1, Accuracy: 0.8725, Time: 0.1003s
Solver: stochastic_gradient_ascent, C: 1.0, Accuracy: 0.73, Time: 0.1388s
Solver: stochastic_gradient_ascent, C: 10.0, Accuracy: 0.6125, Time: 0.1388s
Solver: newton, C: 0.001, Accuracy: 0.8575, Time: 34.0000s
Solver: newton, C: 0.01, Accuracy: 0.86, Time: 25.7885s
Solver: newton, C: 0.1, Accuracy: 0.86, Time: 24.1027s
Solver: newton, C: 1.0, Accuracy: 0.8625, Time: 25.1131s
Solver: newton, C: 10.0, Accuracy: 0.86, Time: 34.9585s


In [ ]:
import plotly.graph_objects as go
import pandas as pd

results_df = pd.DataFrame(results)

fig = go.Figure()

for solver in solvers:
    solver_df = results_df[results_df['solver'] == solver]
    fig.add_trace(go.Scatter(x=solver_df['C'], y=solver_df['accuracy'], mode='lines+markers', name=solver))

fig.update_layout(title='Accuracy vs. C for Different Solvers on l2',
                  xaxis_title='C',
                  yaxis_title='Accuracy',
                  xaxis_type='log')

fig.show()

#Training time plot
fig2 = go.Figure()

for solver in solvers:
    solver_df = results_df[results_df['solver'] == solver]
    fig2.add_trace(go.Scatter(x=solver_df['C'], y=solver_df['training_time'], mode='lines+markers', name=solver))

fig2.update_layout(title='Training Time vs. C for Different Solvers on l2', xaxis_title='C', yaxis_title='Training Time (s)', xaxis_type='log')
fig2.show()



Looking at the performace of accuracy and training time between the three:

###Steepest ascent
Steepest ascent has our best accuracy of .89 with C = .01 and 1. It also generally seems to perform the best out of the three procedures, only falling behind in accuracy at the lowest and middle range with C at .001 or .01. The time taken to train was generally along the same length as Stochastic, only taking a bit longer with the highest C and therefore the least penalizing regulation.

###Stochastic gradient
Stochastic gradient notably suffers with particularly high or low Cs, sometimes ouright falling appart at the steeper ends. It doesn't seem to fare well with heavy or light regularization, and only surpasses other optimizations at a C of 0.1. It's training time matches up with Steepest ascent and is quite low thanks to mini batching speeding up the process.

###Newton
Newton faces the least alterations per C value, staying at around a .86 accuracy for each C. Regularization doesn't seem to have a large effect on it's output largely due to the use of the hessian function. It's hessian calculation is also determining it's very long training time, setting it far apart from the other two optimizations


###Conclusion
Since training times are relatively similiar between the lower two, and steepest ascent's avergae accuracy is more reliable, we will now compare different regularization techniques to find our best performance.

In [ ]:

results = []

for regularization in ['l1', 'l2', 'both']:
  for C in C_vals:
      lr = MultiClassLogisticRegression(eta=0.1, solver = "steepest_ascent", max_iter=500, C=C, regularization=regularization)
      start = time.time()
      lr.fit(X_train, y_train)
      train_time = time.time() - start
      yhat = lr.predict(X_test)
      accuracy = accuracy_score(y_test, yhat)
      results.append({'regularization': regularization, 'C': C, 'accuracy': accuracy, 'training_time': train_time})

      print(f'regularization: {regularization}, C: {C}, Accuracy: {accuracy}, training time: {train_time:.4f}s')

regularization: l1, C: 0.001, Accuracy: 0.84, training time: 0.0891s
regularization: l1, C: 0.01, Accuracy: 0.8225, training time: 0.1052s
regularization: l1, C: 0.1, Accuracy: 0.8875, training time: 0.0861s
regularization: l1, C: 1.0, Accuracy: 0.825, training time: 0.0979s
regularization: l1, C: 10.0, Accuracy: 0.8125, training time: 0.0886s
regularization: l2, C: 0.001, Accuracy: 0.83, training time: 0.1123s
regularization: l2, C: 0.01, Accuracy: 0.89, training time: 0.0871s
regularization: l2, C: 0.1, Accuracy: 0.83, training time: 0.0897s
regularization: l2, C: 1.0, Accuracy: 0.8925, training time: 0.0928s
regularization: l2, C: 10.0, Accuracy: 0.87, training time: 0.0878s
regularization: both, C: 0.001, Accuracy: 0.8225, training time: 0.0956s
regularization: both, C: 0.01, Accuracy: 0.8875, training time: 0.0958s
regularization: both, C: 0.1, Accuracy: 0.83, training time: 0.1083s
regularization: both, C: 1.0, Accuracy: 0.815, training time: 0.0893s
regularization: both, C: 10.0

In [ ]:
results_df = pd.DataFrame(results)

fig = go.Figure()

for reg in ['l1', 'l2', 'both']:
    reg_df = results_df[results_df['regularization'] == reg]
    fig.add_trace(go.Scatter(x=reg_df['C'], y=reg_df['accuracy'], mode='lines+markers', name=reg))

fig.update_layout(title='Accuracy of Steepest Ascent',
                  xaxis_title='C',
                  yaxis_title='Accuracy',
                  xaxis_type='log')

fig.show()

#training time plot

fig2 = go.Figure()

for reg in ['l1', 'l2', 'both']:
    reg_df = results_df[results_df['regularization'] == reg]
    fig2.add_trace(go.Scatter(x=reg_df['C'], y=reg_df['training_time'], mode='lines+markers', name=reg))

fig2.update_layout(title='Training Time of Steepest Ascent', xaxis_title='C', yaxis_title='Training Time (s)', xaxis_type='log')
fig2.show()


After comparing between using l1, l2, and both, it seems that l2 at a C of 1.0 is still the most accurate option. L1's results peak at an average regularization of 0.1 and seem to worsen off the higher C gets. Using both is most similiar to l2 but immediately starts to dovetail at higher values past 0.1 . Training time is negligble here but l2 is generally the fastest, only facing higher rates when C is at it's lowest.

As such, we have chosen Steepest Ascent with a C value of 1.0 using l2 regularization

##Note about Paramter usage (data snooping)

Our parameter selection does involve the evaluation of C and the solver on the test set itself. This can lead to an idealistic estimation of general performance and is called data snooping. We acknowledge that a more proper approach would be to cross validate the training set and ensure that all test data is unused until the model must be evaluated.

##Compare with scklearn
now let's compare this performance with sklearn

In [ ]:
import time
from sklearn.linear_model import LogisticRegression as SKLogisticRegression


start = time.time()
lr_best = MultiClassLogisticRegression(eta=0.1, max_iter=500, solver='steepest_ascent', C=1.0, regularization='l2')
lr_best.fit(X_train, y_train)
custom_time = time.time() - start
yhat = lr_best.predict(X_test)

from sklearn.linear_model import LogisticRegression as SKLogisticRegression
start = time.time()
lr_sk = SKLogisticRegression(C=1.0, penalty='l2', max_iter=1000, solver='liblinear')
lr_sk.fit(X_train, y_train)
sk_time = time.time() - start
yhat_sk = lr_sk.predict(X_test)

comparison = pd.DataFrame([
    {'model': 'Custom (Steepest Ascent)', 'accuracy': accuracy_score(y_test, yhat), 'training_time': custom_time},
    {'model': 'Sklearn (liblinear)', 'accuracy': accuracy_score(y_test, yhat_sk), 'training_time': sk_time}
])
print(comparison)



                      model  accuracy  training_time
0  Custom (Steepest Ascent)    0.8925       0.184481
1       Sklearn (liblinear)    0.9425       0.035899


In [ ]:
# plot accuracy
fig = go.Figure(go.Bar(x=comparison['model'], y=comparison['accuracy']))
fig.update_layout(title='Accuracy: Custom vs Sklearn', yaxis_title='Accuracy', yaxis=dict(range=[0.8, 1.0]))
fig.show()

# plot training time
fig2 = go.Figure(go.Bar(x=comparison['model'], y=comparison['training_time']))
fig2.update_layout(title='Training Time: Custom vs Sklearn', yaxis_title='Seconds')
fig2.show()

Sklearn is generally more accurate and notably faster than our custom logistic regression. Initially, with sklearn's solver at lbfgs the accuracy was marginally higher at around .97 with a significantly increased training time of 5 seconds. This solver was unable to converge with the amount of iterations given, likely because our dataset is unscaled. Liblinear is better suited for unscaled data so we changed the solver to fix the issue.

Liblinear's tighter performance is likely due to it's sophisticated coordinate descent, where it's convergence is handled by optimizing one weight at a time. Steepest ascent updates all weights at once, leading to a less efficient convergence which likely factors into the difference in time and accuracy seen here.



##Deployment


For deployable learning models, I would advise for methods like Scikit learn to be used over a custom logistic regression implementation. Tools like Scikit learn were made with very sophisticated optimizations that improve performance, accuracy, stability, and scalability. They already have built in biases, regularization techniques, and convergences that can. handle overfitting with ease. Additionaly, they come with well documentation and community support. It is much easier for custom implementation to fall short within any of these categories, and is generally not worth the risk of being comparably inefficient. As corroborated by the results of this analysis, Scikit-learn is the ideal choice for deployment.

## Exceptional Anaylsis- One vs One

In [ ]:
from sklearn.multiclass import OneVsOneClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report
import pandas as pd

base_model = LogisticRegression(solver='lbfgs', C=1.0, max_iter=200)

OvO_model = OneVsOneClassifier(base_model)

OvO_model.fit(X_train, y_train)

y_prediction = OvO_model.predict(X_test)

accuracy = accuracy_score(y_test, y_prediction)

report = classification_report(y_test, y_prediction)
print(f"One vs One classification accuracy: {accuracy:.4f}")
print("\nClassification Report: \n", report)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning:

lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression



One vs One classification accuracy: 0.9675

Classification Report: 
               precision    recall  f1-score   support

           1       0.99      0.99      0.99       203
           2       0.97      0.95      0.96       165
           3       0.85      0.91      0.88        32

    accuracy                           0.97       400
   macro avg       0.94      0.95      0.94       400
weighted avg       0.97      0.97      0.97       400



/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning:

lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression



The one v one model was very sucessfull in predicting the burnout labels. It was most sucessful in predicting the 1st label, and the least in predicting the 3rd label (but still pretty good, 85% precision). Overall, the model has an accuracy of 96%. This is better than our custom steepest ascent model that had an accuracy of 89%, and the sklearn model 94%. OvO performs well on datasets that have a high class imbalance, which makes sense considering that our dataset has a major imbalance in the 1/2 classes vs the 3 class.